lib/compat/wordpress-5.9/class-wp-theme-json-5-9.php

```diff
@@ -69,79 +69,93 @@ class WP_Theme_JSON_5_9 {
 	 *
 	 * This contains the necessary metadata to process them:
 	 *
-	 * - path       => where to find the preset within the settings section
-	 * - override   => whether a theme preset with the same slug as a default preset
-	 *                 can override it
-	 * - value_key  => the key that represents the value
-	 * - value_func => optionally, instead of value_key, a function to generate
-	 *                 the value that takes a preset as an argument
-	 *                 (either value_key or value_func should be present)
-	 * - css_vars   => template string to use in generating the CSS Custom Property.
-	 *                 Example output: "--wp--preset--duotone--blue: <value>" will generate as many CSS Custom Properties as presets defined
-	 *                 substituting the $slug for the slug's value for each preset value.
-	 * - classes    => array containing a structure with the classes to
-	 *                 generate for the presets, where for each array item
-	 *                 the key is the class name and the value the property name.
-	 *                 The "$slug" substring will be replaced by the slug of each preset.
-	 *                 For example:
-	 *                 'classes' => array(
-	 *                   '.has-$slug-color'            => 'color',
-	 *                   '.has-$slug-background-color' => 'background-color',
-	 *                   '.has-$slug-border-color'     => 'border-color',
-	 *                 )
-	 * - properties => array of CSS properties to be used by kses to
-	 *                 validate the content of each preset
-	 *                 by means of the remove_insecure_properties method.
+	 * - path             => Where to find the preset within the settings section.
+	 * - prevent_override => Whether a theme preset with the same slug as a default preset
+	 *                       should not override it or the path to a setting for the same
+	 *                       When defaults.
+	 *                       The relationship between whether to override the defaults
+	 *                       and whether the defaults are enabled is inverse:
+	 *                         - If defaults are enabled  => theme presets should not be overriden
+	 *                         - If defaults are disabled => theme presets should be overriden
+	 *                       For example, a theme sets defaultPalette to false,
+	 *                       making the default palette hidden from the user.
+	 *                       In that case, we want all the theme presets to be present,
+	 *                       so they should override the defaults by setting this false.
```

Comment:
are we introducing this concept, `prevent_override` here?
is there a way to reframe it in the assertive positive sense? I always have trouble understanding contradictions or negative constructions. for example, instead of `prevent_override = true` maybe `hide_defaults = false`?

not that `hide_defaults` is certainly what we want, but it's actively what we're doing (or maybe "hide" is wrong, but we're dong something), instead of what we're not doing.

Reply:
> Are we introducing this concept, prevent_override here? Is there a way to reframe it in the assertive positive sense? I always have trouble understanding contradictions or negative constructions.

It was kind of like that before, but I changed it hoping that it would make things less confusing despite being a negative construction. I'll explain my thought process for replacing `override` with `prevent_override`.

Arrays in the `PRESETS_METADATA` reference values in theme.json, so let's take this theme.json as an example. I'm using `defaultPalette` here, but `defaultGradients` and `defaultDuotone` should behave the same way.

```json
{
  "settings": {
    "color": {
      "defaultPalette": false
    }
  }
}
```

Previously, there was a key called `override` that could be a path array or a boolean. 

```php
const PRESETS_METADATA = array(
  array(
    'override' => array( 'color', 'defaultPalette' ), // Path to a value in theme.json settings
  ),
  array(
    'override' => true, // A hard-coded value that doesn't have a corresponding theme.json setting
  ),
);
```

I expected the values to resolve like this when I was reading it.

```php
const PRESETS_METADATA = array(
  array(
    'override' => false // Substitute in the value from theme.json
  ),
  array(
    'override' => true // Pass through a boolean
  ),
);
```

But actually, there was a function, [`should_override_preset`](https://github.com/WordPress/gutenberg/blob/768a8e61e1c3ead692647b70c7cbe6cd7cbbe57c/lib/compat/wordpress-5.9/class-wp-theme-json-5-9.php#L1542-L1579), that would look up the value and _invert_ it or just pass through a boolean _without_ inverting it. So the actual substitution would look like this.

```php
const PRESETS_METADATA = array(
  array(
    'override' => true // Substitute in the value from theme.json _and_ invert it
  ),
  array(
    'override' => true // Booleans don't get inverted
  ),
)
```

As part of this PR I added `use_default_presets` which also looks up a value from the theme.json, but does _not_ need to invert the lookup. 

```php
const PRESETS_METADATA = array(
  array(
    'use_default_presets' => array( 'color', 'defaultPalette' ),
  ),
  array(
    'use_default_presets' => true,
  ),
);
```

It needs to resolve to this.

```php
const PRESETS_METADATA = array(
  array(
    'use_default_presets' => false // Substitute in the value from theme.json
  ),
  array(
    'use_default_presets' => true // Pass through a boolean
  ),
);
```

Between the weird behavior of `should_override_preset` and wanting consistency for how theme.json settings are looked up, it made more sense to me to replace `override` with `prevent_override` and use a more generic `get_metadata_boolean` for both of them.

Does that make sense to you? Would you rather have inconsistent path array meanings or a negative construction? Or is there another simple option that I'm overlooking?

Reply:
it's too bad the JSON file has the inversion, rather than something like, `showDefaultPallette` or `useDefaultPallette` or `includeDefaultPalette`

probably the confusion stems from having multiple speakers saying the same words but from different perspectives.

should `prevent_override` protect the default values, the hard-coded values, or the JSON file values?

in the `PRESETS_METADATA` examples are those priority lists? or are those two separate settings, such as `palette` and `gradient` etc…?

I guess the idea is that if we have `true` or `false` as the value of `use_default_presets` or for `prevent_override` then that value will be used int he end. if we set a path as an array for the field though it will read from the JSON and use that?

what is the relationship between `prevent_override` and `use_default_presets`? pardon my confusion; I'm trying to learn something about this system but still feel ignorant. is there a logic table with the inputs and desired outcomes for this?

My confusion peaks in lines like these. It's hard for me to read through the code and explain what impact `$skip(_default_presets)` is supposed to have on the end result.

```php
$skip = ! static::get_metadata_boolean( $this->theme_json['settings'], false, true );
$skip = ! static::get_metadata_boolean( $this->theme_json['settings'], true, true );
$skip = ! static::get_metadata_boolean( $this->theme_json['settings'], [ 'color', $skip = 'defaultPalette' ], true );
```

Are "defaults" and "presets" different?

Reply:
Lots of good questions! I'll answer them a bit out of order in a way that they can build upon one another.

---

> Are "defaults" and "presets" different?

A preset is the metadata needed to generate the relevant global styles parts: CSS variables for everything, CSS classes for things like text/background color, and SVGs for things like duotone filters.

They come from three different origins listed here in the order that they are processed:
- "default" presets are hard-coded in lib/theme.json. They were previously called "core" presets.
- "theme" presets come from a block theme's theme.json file.
- "custom" presets come from the styles sidebar of the site editor. They were previously called "user" presets.

---

> Should `prevent_override` protect the default values, the hard-coded values, or the JSON file values?

`prevent_override` protects the slugs used by default presets.

So maybe we could name it `protect_default_slugs` or `reserve_default_slugs` or something? Would that make more sense?

The `override` option was added in #36811 and updated to use a path in #37008. tl;dr It was added to fix the UI by skipping theme slugs that already existed in the default palette. Theme developers that had `"defaultPalette": false` in their theme expected to be able to use these reserved default slugs, so it was then tied to `defaultPalette`.

These screenshots from #36811 illustrate it well—`black` and `white` are duplicate keys in TT1-blocks and were colored hot pink for emphasis.

> | Before | After |
> | --------- | ------- |
> | ![Captura de ecrã de 2021-11-24 10-35-44](https://user-images.githubusercontent.com/583546/143213063-2bfb5a46-ed29-4e20-9bdf-5bee4f45e5c0.png) |  ![Captura de ecrã de 2021-11-24 10-38-16](https://user-images.githubusercontent.com/583546/143213179-f7b82479-7944-47a5-9b39-296883619718.png) |

Let's run through an example with a slightly modified theme.json from the #36811 example to see what is being generated from the presets.

```jsonc
{
  "settings": {
    "color": {
      "defaultPalette": true,
      "palette": [
        {
          "slug": "black",
          "color": "#FF69B4"
        },
        {
          "slug": "dark-gray",
          "color": "#2A303C"
        },
        // ...other theme preset colors...
        {
          "slug": "white",
          "color": "#FF69B4"
        }
      ]
    }
  }
}
```

Here are the relevant CSS variables that are generated.

```css
body {
  --wp--preset--color--black: #000000;
  --wp--preset--color--cyan-bluish-gray: #abb8c3;
  --wp--preset--color--white: #ffffff;
  /* ...other default preset colors... */
  --wp--preset--color--dark-gray: #2A303C;
  /* ...other theme preset colors... */
}
```

Notice that the `white` and `black` values we defined in theme.json do not get generated. The value for `--wp--preset--color--white` is `#ffffff` instead of `#FF69B4`. When `defaultPalette` is `true`, the default colors take precedence. 

My expectation as a block theme creator would be that values I pass replace the default ones, but the expected outcome according to #36811 and #37008 is the other way around.

---

> What is the relationship between `prevent_override` and `use_default_presets`?

`prevent_override` controls how theme presets with the same slug as default presets should be handled as explained above.

`use_default_presets` was added by this PR to wholesale prevent generating anything from default presets (including the SVGs in question) when its value is `false`. [My take](https://github.com/WordPress/gutenberg/pull/38681#discussion_r803119718) was that the color/gradient/duotone palettes should all behave the same, so I fixed the default generation for them as well.

In the case of the color/gradient/duotone palettes. I suppose you could say that `use_default_presets` obsoletes `prevent_override`.

However, in the case of `fontSizes` and `fontFamilies`, the values differ. The default values should always generate _and_ the theme preset slugs should override the default preset slugs like they did before.

> Is there a logic table with the inputs and desired outcomes for this?

| `use_default_presets` | `prevent_override` | outcome |
| --------------------- | ------------------ | ------- |
| `false`               | N/A                | Nothing is generated from default presets |
| `true`                | `false`            | Generate default presets, but replace the value when the theme preset slug matches the default preset slug (theme wins) |
| `true`                | `true`             | Generate default presets, but skip the theme preset when the slug matches the default preset slug (default wins) |

---

> In the `PRESETS_METADATA` examples are those priority lists? Or are those two separate settings, such as palette and gradient etc…?

They are separate settings. You could add duplicates to the list, and I expect it would generate duplicates in the output.

However, `ALLOWED_ORIGINS` is kind of a priority list with the exceptions noted above for the `default{Palette,Gradients,Duotone}` option in theme.json.

---

> I guess the idea is that if we have `true` or `false` as the value of `use_default_presets` or for `prevent_override` then that value will be used in the end. if we set a path as an array for the field though it will read from the JSON and use that?

Exactly.

Reply:
that's extremely helpful and clarifying @ajlende, thank you for taking the effort to add that.

> They were previously called "core" presets.…They were previously called "user" presets.

Oh no! those names make so much more sense to me 😆. I trust there was a good reason to diverge from them.

---

please don't let this derail the PR or the fixes, it's me doodling on the idea.

it's almost like we're looking at a natural hierarchy and talking about what behavior happens when finding duplicate values. the order will always be `core` (default) applies first, `theme` second, and `user` third. at any point in the hierarchy we can choose to ignore a new duplicated value and let the previous assignment remain _or_ replace the previous assignment with our new value.

the kicker is that sometimes we don't want `core` values enter into the equation at all. we want a blank slate.

❓ is there any reason a `custom` config couldn't start clean as well and wipe out all of the previously-assigned values?

```ts
{
	// sets completely new duotone values regardless
	// of what was previously set
	"color": {
		"resetConfig": true,
		"duotone": { … }
	}
}

{
	// retains existing values but if we find a new
	// value for the same key, use the new one
	"color": {
		"duplicateKeys": "replace",
		"duotone": { … }
	}
}

{
	// retains existing values but if we find a new
	// value for the same key, use the old one
	"color": {
		"duplicateKeys": "retain",
		"duotone": { … }
	}
}
```

I wonder if this description could account not only for `use_default` but also `prevent_override` and make it work at both the theme and the custom levels. a theme could decide to `resetConfig` and the custom level could extend that.

on the other hand, it couldn't bring back the default levels with this description since the theme might have already reset them.

---

my other idea was wondering if we are standardized on `default`, `theme`, and `custom`, if we _could_ go ahead and set the priority.

```ts
{
	// removes default config values entirely,
	// custom config values take priority over theme
	"color": {
		"configs": [ "theme", "custom" ],
		"duplicateKeys": "replace",
		"duotone": { … }
	}
}
```

while a bit more verbose this would give each level in the hierarchy the control to decide which values should appear and which should take precedence in a conflict. if a theme turned off the default values then a custom config could turn them back on. a custom config to let the defaults stay but wipe out the theme's colors.

any thoughts? way too much discussion about something that's already embedded in the system? my noodling is mostly around thinking through what exactly the behaviors are we're doing and avoiding special-casing in the interface, such as when calling out `default` in `use_default`.

Reply:
> Any thoughts? Way too much discussion about something that's already embedded in the system?

I'm happy to explain the existing system and how it drove my decisions for this bug fix, but, yeah, we're starting to get into big picture changes with some of these. I'm just trying to make sure SVG filters don't get left behind when changes are made to colors/gradients/etc. and, in this case, fixing a bug that is more obvious because of SVG filters.

Still, these questions might be interesting for @oandregal and/or @jorgefilipecosta to see since they are more focused on global styles than I am.